### ISSUE TITLE
List priority parameter not properly used to prioritize incoming spool

### ISSUE THREAD

**Issue by salaun-urennes1** [sympa-community/sympa#1376]
I recently realized that list `priority` parameter is not properly used to prioritize incoming spool. We had our incoming `msg/` spool plenty of 4000 automatically generated mails. It prevented processing of mails for other mailing lists. I tried changing that list priority from 5 to 9 and restarted the `sympa_msg.pl` process. I noticed that despite this priority change sympa_msg.pl would go on processing mails with the lower priority. 

Version
-------
6.2.68

Installation method
-------------------
From sources

Expected behavior
-----------------
`sympa_msg.pl` should process mails with higher priority first, regardless of the file creation time in the spool.

Actual behavior
---------------
`sympa_msg.pl` processes the older files in spool first.

Steps to reproduce
------------------
Create two listes : `higher_priority` and `lower_priority`.
Configure `lower_priority` list with priority 9 (default being 5).
Stop `sympa_msg.pl` process
Send an email to `lower_priority` first
Then send an email to `higher_priority` list.
Start `sympa_msg.pl` process
=> you will notice that mail for `lower_priority` list will be processed first.

Additional information
----------------------
Attached is a patch to `Sympa/Spool/Incoming.pm`. It changes the `_load()` method behavior to order the spool content with an additional criteria related to priority. I tested that patch and applied it on our production server. It would probably also make sense to adapt `_filter()` method too.
[20220413-incoming-priority.patch.txt](https://github.com/sympa-community/sympa/files/8479888/20220413-incoming-priority.patch.txt)

**Comment by salaun-urennes1** [comment]
If we want the incoming spool processing to be even more dynamic regarding mail priority, it would make sense to reload the spool content before the end of its processing, just in case new files with higher priority have been created in the meanwhile

**Comment by ikedas** [comment]
### Physical limits

The limits of specification we assume must be taken into account.  For the size of list: According to [README](https://sympa-community.github.io/sympa/), "a list with 20 000 subscribers" looks ordinary.  On the contrary it is reported that there is a list with approximately 10^7 subscribers (I can't remember where I read it). --- The number of messages stored in the spool at once is also considered to be around the same.

Under this assumption, the code to hold the unmarshalled metadata for all messages in memory at once is unacceptable.  Because it will require unrealistically large amount of physical memory.

That's why `Sympa::Spool::_load()` (or the same method of deriving class) never unmarshall the metadata of messages in spool: It is allowed to compare marshalled metadata (filename) only.  Unmarshalling is done during `Sympa::Spindle::spin()` (or the same), and moreover, for the same reason, here is always only one message unmarshalled at a time: all messages in the spool are never unmarshalled at once.  Therefore, explosion of memory usage will be avoided.


### Current algorithm: Lazy prioritisation

By the way, currently prioritisation of messages in incoming (`msg`) spool is not done by `Sympa::Spool::Incoming` but by [`Sympa::Spindle::ProcessIncoming`](https://sympa-community.github.io/gpldoc/man/Sympa-Spindle-ProcessIncoming.3.html):

> Order to process messages in source spool are controlled by modification time
> of files and delivery date.
> Some messages are skipped according to these priorities
> (See [Sympa::Spool::Incoming](https://sympa-community.github.io/gpldoc/man/Sympa-Spool-Incoming.3.html)):
> 
> - Messages with lowest priority (`z` or `Z`) are skipped.
> - Messages with possibly higher priority are chosen.
> This is done by skipping messages with lower priority than those already
> found.

See the code for details of processing.  This means that --- Even if there are higher priority messages in the spool, some of the lower priority messages may be processed first (especially if they are older).  However, if there are sufficiently large number of messages in the spool, the overall tendency will be to process more high priority messages.

One of the reasons for such an algorithm is that, due to the limitation described in the previous section (not all metadata can be loaded into memory at once), it is not possible to sort all messages in metadata order.  However, in some situations this algorithm in itself is not so bad.

If it was decided to process all the messages in the spool that have a high priority before processing the lower ones,  it may probably lead to undesirable behavior:

  * Low priority messages are blocked by high priority messages. For example: While some lists have problems and a large number of listmaster notifications (highest priority) are being sent, delivery of the other lists may be stopped; Furthermore, if list postings are active, the response for the mail command may be stopped.

  * The frequency of message delivery tends to be intermittent, which reduces the efficiency of network utilisation.

As above, the algorithm "simply prioritise high priority messages" may not be optimal in all situations (at least, no particular benefit can be derived in situations where there are only a few messages in the spool: Messages in the spool are ejected as soon as possible, so time-ordered processing is more suitable).

What algorithm is really optimal should be determined by detailed verification, e.g. based on careful simulations. At the very least, I do not consider the current algorithm to be grossly inadequate.

**Comment by salaun-urennes1** [comment]
First of all, thank you for such a detailed and argumented response.
My comments below:

> The limits of specification we assume must be taken into account. For the size of list: According to [README](https://sympa-community.github.io/sympa/), "a list with 20 000 subscribers" looks ordinary. On the contrary it is reported that there is a list with approximately a few 10^5 subscribers (I can't remember where I read it). --- The number of messages stored in the spool at once is also considered to be around the same.
I don't think you can compare the list size with the number of messages in the spool. If I take our mailing lists service as an example: we have 6.200 lists, splitted in 11 robots. Our biggest list has 41.000 members. And most of the time our incoming spool is empty...hopefully. When it sometimes grows (because of mailing lists receiving automated notifications) , I just don't want to penalize the most important mailing lists used by students, teachers or administration of the University.

> Under this assumption, the code to hold the unmarshalled metadata for all messages in memory at once is unacceptable. Because it will require unrealistically large amount of physical memory.
I understand your concern, but It seems to me that, at the end of the spool processing you will also have all unmarshalled metadata in memory. Currently unmarshalling is performed by `Sympa::Spool::next()`, ran on each file in the spool, but one at a time. `Sympa::Spool::unmarshal_metadata()` collects some informations (`listname`, `listtype`, `priority`), but the biggest piece of information seems to be `$data->{context}` because it contains the whole `List` object (if the file in spool concerns a list). However Sympa::List::new() keeps a cache of List objects in memory (`%list_of_lists`). So regarding memory consumption, it seems that the consuption is the same if you run `Sympa::Spool::unmarshal_metadata()` at once (my proposed patch) or if you run it on each message until the end of the spool (current code).

> ### Current algorithm: Lazy prioritisation
> 
> By the way, currently prioritisation of messages in incoming (`msg`) spool is not done by `Sympa::Spool::Incoming` but by [`Sympa::Spindle::ProcessIncoming`](https://sympa-community.github.io/gpldoc/man/Sympa-Spindle-ProcessIncoming.3.html):
Thank you for this information. However, I observe that prioritization of messages based on `mtime` and `priority` parameter is treated at the `Sympa::Spool` level.

> > * Messages with lowest priority (`z` or `Z`) are skipped.
> > * Messages with possibly higher priority are chosen.
> >   This is done by skipping messages with lower priority than those already
> >   found.
> 
> See the code for details of processing. This means that --- Even if there are higher priority messages in the spool, some of the lower priority messages may be processed first (especially if they are older). However, if there are sufficiently large number of messages in the spool, the overall tendency will be to process more high priority messages.
I checked that code and because I considered it inappropriate I proposed the patch.

Maybe it's worth developing a scenario: 
- if you have a "healthy" Sympa server you should have your incoming spool empty most of the time,
- in our case it grew (up to 4.000 mails) because bulk notifications had been sent to our monitoring mailing list by automated process. We ended up with 4.000 files for the lower_priority list in incoming spool + 50 more recent files for higher_priority lists. 
- no chance any of the 50 files would be processed before the 4.000 files, even if you restart `sympa_msg.pl` process, because files priority is not considered globally in the spool, but one file after the other.

...
> What algorithm is really optimal should be determined by detailed verification, e.g. based on careful simulations. At the very least, I do not consider the current algorithm to be grossly inadequate.
I can spend some time in simulations, if it can help...

**Comment by ikedas** [comment]
> First of all, thank you for such a detailed and argumented response. My comments below:
> 
> > The limits of specification we assume must be taken into account. For the size of list: According to [README](https://sympa-community.github.io/sympa/), "a list with 20 000 subscribers" looks ordinary. On the contrary it is reported that there is a list with approximately a few 10^5 subscribers (I can't remember where I read it). --- The number of messages stored in the spool at once is also considered to be around the same.
> 
> I don't think you can compare the list size with the number of messages in the spool. If I take our mailing lists service as an example: we have 6.200 lists, splitted in 11 robots. Our biggest list has 41.000 members. And most of the time our incoming spool is empty...hopefully. When it sometimes grows (because of mailing lists receiving automated notifications) , I just don't want to penalize the most important mailing lists used by students, teachers or administration of the University.

Would my estimate be excessive? But in any case, I think we must consider the worst logically possible limits rather than examples from actual experience (as an example, I once experienced 7,000 lists in single domain.  But logically more lists may be possible).

> > Under this assumption, the code to hold the unmarshalled metadata for all messages in memory at once is unacceptable. Because it will require unrealistically large amount of physical memory.
> 
> I understand your concern, but It seems to me that, at the end of the spool processing you will also have all unmarshalled metadata in memory. Currently unmarshalling is performed by `Sympa::Spool::next()`, ran on each file in the spool, but one at a time. `Sympa::Spool::unmarshal_metadata()` collects some informations (`listname`, `listtype`, `priority`), but the biggest piece of information seems to be `$data->{context}` because it contains the whole `List` object (if the file in spool concerns a list). However Sympa::List::new() keeps a cache of List objects in memory (`%list_of_lists`). So regarding memory consumption, it seems that the consuption is the same if you run `Sympa::Spool::unmarshal_metadata()` at once (my proposed patch) or if you run it on each message until the end of the spool (current code).

Unmarshalling each metadata generates a hash reference containing metadata which consumes much more memory than single string (marshalled filename).  If a large number of metadata are held, the memory consumption (and also time for processing) may be non-negligible.

> > ### Current algorithm: Lazy prioritisation
> > By the way, currently prioritisation of messages in incoming (`msg`) spool is not done by `Sympa::Spool::Incoming` but by [`Sympa::Spindle::ProcessIncoming`](https://sympa-community.github.io/gpldoc/man/Sympa-Spindle-ProcessIncoming.3.html):
> 
> Thank you for this information. However, I observe that prioritization of messages based on `mtime` and `priority` parameter is treated at the `Sympa::Spool` level.
> 
> > > * Messages with lowest priority (`z` or `Z`) are skipped.
> > > * Messages with possibly higher priority are chosen.
> > >   This is done by skipping messages with lower priority than those already
> > >   found.
> > 
> > 
> > See the code for details of processing. This means that --- Even if there are higher priority messages in the spool, some of the lower priority messages may be processed first (especially if they are older). However, if there are sufficiently large number of messages in the spool, the overall tendency will be to process more high priority messages.
> 
> I checked that code and because I considered it inappropriate I proposed the patch.
> 
> Maybe it's worth developing a scenario:
> 
> * if you have a "healthy" Sympa server you should have your incoming spool empty most of the time,
> * in our case it grew (up to 4.000 mails) because bulk notifications had been sent to our monitoring mailing list by automated process. We ended up with 4.000 files for the lower_priority list in incoming spool + 50 more recent files for higher_priority lists.
> * no chance any of the 50 files would be processed before the 4.000 files, even if you restart `sympa_msg.pl` process, because files priority is not considered globally in the spool, but one file after the other.
> 
> ...
> 
> > What algorithm is really optimal should be determined by detailed verification, e.g. based on careful simulations. At the very least, I do not consider the current algorithm to be grossly inadequate.
> 
> I can spend some time in simulations, if it can help...

That would be a very significant work. It would be greatly appreciated if simulations could be made for various situations and even the impact on outgoing deliveries etc. taken into account.

**Comment by ikedas** [comment]
I have an idea and leave it as a comment.

Here is a summary of priorities for the messages thrown into incoming spool (i.e. `msg` spool).

| Address of recipient  | parameter          | default |
| ----------------------------- | ------------------ | ------- |
| `listmaster`                  | (hardcoded)        | `0` (fixed value) |
| `*-request` (owner address)   | `request_priority` | `0`     |
| `sympa`  (address for mail command)  | `sympa_priority`   | `1`     |
| the other (list post address) | `priority` (by each list) / `default_priority` | `5` |

According to above, it looks to me that the need to dynamically change priorities using parameters seems scarce.

In other words, the parameter could be deprecated and priorities could be set with the queue program (as those in early versions) such as below.  This way, there would be no increase in memory usage and no need to add complex exception handling to the spool code.

``` code
listmaster:   "|/path/to/queue 0 sympa@my.dom.ain"
sympa:        "|/path/to/queue 1 sympa@my.dom.ain"
list-request: "|/path/to/queue 0 sympa@my.dom.ain"
list:         "|/path/to/queue 5 sympa@my.dom.ain"
```
This need only a slight changes:
``` diff
diff --git a/src/libexec/queue.c b/src/libexec/queue.c
index 1e1e924..f65bbb4 100644
--- a/src/libexec/queue.c
+++ b/src/libexec/queue.c
@@ -93,6 +93,7 @@ main(int argn, char **argv)
    char        *queuedir;
    char        *listname;
    int                 firstline = 1;
+   char priority;

    /* Usage : queue list-name */
    if ((argn < 2) || (argn >3)) {
@@ -102,13 +103,19 @@ main(int argn, char **argv)
    }

    if (argn == 2) {
+      priority = '5';
       listname = malloc(strlen(argv[1]) + 1);
       if (listname != NULL)
          strcpy(listname, argv[1]);
    }

-   /* Old style : queue priority list-name */
+   /* Alternative style : queue priority list-name */
    if (argn == 3) {
+      if (argv[1][0] && !argv[1][1])
+         priority = argv[1][0]);
+      else
+         priority = '5';
+
       listname = malloc(strlen(argv[2]) + 1);
       if (listname != NULL)
          strcpy(listname, argv[2]);
@@ -126,7 +133,7 @@ main(int argn, char **argv)
      exit(EX_NOPERM);
    }
    umask(027);
-   snprintf(qfile, sizeof(qfile), "T.%s.%ld.%d", listname, time(NULL), getpid());
+   snprintf(qfile, sizeof(qfile), "T.%c,%s.%ld.%d", priority, listname, time(NULL), getpid());
    fd = open(qfile, O_CREAT|O_WRONLY, 0600);
    if (fd == -1){
      char* buffer=(char*)malloc(strlen(argv[0])+strlen(queuedir)+80);
```

**Comment by ikedas** [comment]
Though I consider it scarce, if you say it is occasionally useful, I don't disagree unreservedly.

One more thing that came to my mind  is that: By multiple sympa_msg workers (configured with `incoming_max_count`), is the situation you illustrate could be mitigated?

**Comment by salaun-urennes1** [comment]
I tried setting `incoming_max_count` to 2, but systemD failed to restart the `sympa_msg` processes:
```
# systemctl restart sympa
Job for sympa.service failed because a timeout was exceeded. See "systemctl status sympa.service" and "journalctl -xe" for details.
```
I had to restore the previous `sympa.conf`

**Comment by ikedas** [comment]
Could you please show the logs of Sympa?

**Comment by salaun-urennes1** [comment]
Here are the sympa logs during the restart :
```
Nov 23 16:33:14 vmperl-psympa1 sympa_msg[2785]: notice main::sigterm() Signal TERM received, still processing current task
Nov 23 16:33:14 vmperl-psympa1 sympa_msg[2785]: notice main:: Sympa/msg exited normally due to signal
Nov 23 16:33:17 vmperl-psympa1 sympa_msg[23671]: info main::_load() Configuration file read, default log level 0
Nov 23 16:33:17 vmperl-psympa1 sympa_msg[23671]: notice Sympa::Process::daemonize() Starting sympa/msg daemon, PID 23674
Nov 23 16:33:17 vmperl-psympa1 sympa_msg[23674]: info main:: Starting child daemon, PID 23675
Nov 23 16:33:17 vmperl-psympa1 sympa_msg[23675]: info main:: Slave daemon started with PID 23675
Nov 23 16:33:17 vmperl-psympa1 sympa_msg[23675]: notice main:: Sympa/msg 6.2.48 Started
Nov 23 16:33:18 vmperl-psympa1 sympa_msg[23674]: notice main:: Sympa/msg 6.2.48 Started
...
Nov 23 16:33:19 vmperl-psympa1 sympa_msg[23675]: notice main::sigterm() Signal TERM received, still processing current task
Nov 23 16:33:19 vmperl-psympa1 sympa_msg[23674]: notice main::sigterm() Signal TERM received, still processing current task
Nov 23 16:33:19 vmperl-psympa1 sympa_msg[23674]: notice main:: Sympa/msg exited normally due to signal
Nov 23 16:33:19 vmperl-psympa1 sympa_msg[23675]: notice main:: Sympa/msg exited normally due to signal
```
It shows that `incoming_max_count` parameter was applied, because 2 `sympa_msg` have been started. The "..." corresponds to normal processing of incoming messages. I don't understand why both `23674` and `23675` received a `TERM` signal; no error in the logs between 16:33:17 and 16:33:19.

**Comment by ikedas** [comment]
`sympa.service` will be stopped if other related service(s) stop.  They may be killed (e.g. by OOM killer) or may simply have failed starting.
